File: 7 - Processamento de Imagens/Atividades Práticas/Atividade01/atividade.m
% Carrega as imagens
filamento = imread('tungsten_filament_shaded.tif');
sombreamento = imread('tungsten_sensor_shading.tif');

% Converte as imagens para double para evitar problemas de aritmética de inteiros
filamento_double = im2double(filamento);
sombreamento_double = im2double(sombreamento);

% Realiza a correção do sombreamento pela divisão da imagem pelo padrão de sombreamento
% Nota: É importante evitar a divisão por zero ou valores muito baixos no padrão de sombreamento.
% Podemos adicionar um pequeno valor (por exemplo, eps) para garantir a estabilidade numérica.
corrigida_double = filamento_double ./ (sombreamento_double + eps);

% Reescalonamento dos valores para o intervalo original de 0 a 255
corrigida = im2uint8(mat2gray(corrigida_double));

% Exibe a imagem original e a corrigida
figure;
subplot(1, 2, 1); imshow(filamento); title('Original');
subplot(1, 2, 2); imshow(corrigida); title('Corrigida');

% Salva a imagem corrigida
imwrite(corrigida, 'tungsten_filament_corrected.tif');

File: 7 - Processamento de Imagens/Atividades Práticas/Atividade01/atividade.py
import cv2
import numpy as np

# Carrega as imagens
filamento = cv2.imread('tungsten_filament_shaded.tif', cv2.IMREAD_UNCHANGED)
sombreamento = cv2.imread('tungsten_sensor_shading.tif', cv2.IMREAD_UNCHANGED)

# Converte as imagens para double para evitar problemas de aritmética de inteiros
filamento_float = filamento.astype(np.float32)
sombreamento_float = sombreamento.astype(np.float32)

# Evita a divisão por zero adicionando um pequeno valor ao denominador
sombreamento_float[sombreamento_float == 0] = np.finfo(np.float32).eps

# Realiza a correção do sombreamento
corrigida_float = filamento_float / sombreamento_float

# Normaliza a imagem corrigida para o intervalo [0, 255] e converte para uint8
corrigida = cv2.normalize(corrigida_float, None, 0, 255, cv2.NORM_MINMAX).astype('uint8')

# Exibe a imagem original e a corrigida
cv2.imshow('Original', filamento)
cv2.imshow('Corrigida', corrigida)
cv2.waitKey(0)
cv2.destroyAllWindows()

# Salva a imagem corrigida
cv2.imwrite('tungsten_filament_corrected.tif', corrigida)
